### ISSUE TITLE
[wpe-2.28] Black screen on lightning apps when visibility is toggled

### ISSUE THREAD

**Issue by varumugam123** [Metrological/WebKitForWayland#882]
With wpe-2.28 (aa569f0366), on comcast devices we observe blackscreen on an active lightning application if visibility is set first to false & then to true. It looks like the application doesn't realize the visibility change and doesn't make attempt to refresh/render the contexts.

The issue was observed on AppleTV+, HBOMax and other apps.

Steps to reproduce the issue:
Launch any of the Lightning apps (with nonCompositedWebGL set to true) and set visibility to false (or hidden to true) once the launch is completed, make sure when you do this there are no animations on the page (you might want to scroll down / up in the app to render static content)... Then restore the visibility (i.e visibility to true)... 

Expected Result : App should render the content on screen

Observation : Black screen until any key input is sent that triggers a content update

Environment : Observed on comcast device but the issue seems generic. The app was launched with WebKitBrowser plugin...

**Comment by magomez** [comment]
This happens because when on nonCompositedWebGL mode, the content is cleared to transparent when the browser is suspended/hidden. When the browser becomes visible/resumed, the application should repaint its content in order to become visible again. This works in general because the application animations will trigger the repaint, but if the app content is static, then the content is not repainted. I'll work on a change in order to save the app content before suspend/hide and restore it during resume/show.

**Comment by magomez** [comment]
This happens on nonCompositedWebGL because there webgl renders directly to the window. When the browser is hidden/suspended, it paints the window transparent, effectively losing the content that the application was painted. When the browser is then shown/resumed, we don't have anything on the window until the application renders something again. And if the application doesn't have animations then it doesn't render anything, so the browser stays transparent. This doesn't happen with nonCompositedWebGL disabled, because in that case the webgl content is painted into an intermediate texture that is then composited, so we have the webgl content saved to paint it again when the browser gets shown/resumed.

I've been working on a patch to store the browser window content when hiding/suspending and paint it back when showing/resuming, but while working on this I realized about a problem: while saving the content is trivial, painting it back requires performing some extra work with opengl. The problem with this is that these operations need to be executed in the glcontext that the page is using through webgl. This means that both the application and the browser are going to be messing with the opengl resources, and when trying to paint the display content back, there's the danger of modifying something that the page is using, causing visual glitches in the application rendering afterwards. Whether these glitches may happen or not depends on how the page is using the resources, so it's not easy to know beforehand.

In the patch I'm working on, (I think) I've been able to avoid many problems by storing the values that the page is using before modifying them, and set the values back after I'm done, but there are some resources where this is not easy to achieve. And even with that, I'm not very happy about this approach to fix the problem.

@woutermeek I think it would be way easier to fix on the lightning side. AFAIK lightning apps monitor whether the browser is visible or not, and they could redraw their content when they become visible, even if there aren't animations ongoing. Could you check whether that's an option with someone from the lightning team? If not, I can continue with this approach.

**Comment by woutermeek** [comment]
@michielvandergeest @erikhaandrikman would this be something to handle in the Lighting SDK?

**Comment by michielvandergeest** [comment]
If I understand correctly we're talking about adding some event listener that triggers the renderer to do a repaint to make the content visible again. Basically automating the behaviour that can now be achieved by a user input or an animation.

Possibly we can add that to the Lightning-SDK yes. I'm just not sure how this visible event is being emitted? Is that a browser event?

**Comment by woutermeek** [comment]
Yes, there is a native event for it.
https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event

**Comment by michielvandergeest** [comment]
Cool! Yeah, we can get this functionality added on a separate branch of the SDK early next week, so we can test if it resolves the problem.

**Comment by woutermeek** [comment]
That would be awesome!

**Comment by michielvandergeest** [comment]
I've pushed a potential fix for this issue to the `fix/visibility-toggle-black-screen`

@magomez are you able to test if this solution indeed solves the issue?

**Comment by woutermeek** [comment]
@varumugam123 Could you check with the updated SDK if that resolves the problem?

**Comment by varumugam123** [comment]
It looks like AppleTV+ is not using Lightning-SDK (it is directly using Lightning Core components). So that change in Lightning-SDK wouldn't solve AppleTV+ issue.

Also, when I tried a sample lightning app built using the steps mentioned in https://lightningjs.io/docs/#/getting-started/CreateApp/index, what I see is the visibility change doesn't cause any issue i.e there is no black screen observed when visibility is set to false and then to true. So something in Lightning-SDK is making up to cover the actual issue.

**Comment by varumugam123** [comment]
I stand corrected... The behavior I observed with the test lightning app (no black screen on visibility change) was because of the animation the app has i.e it periodically refreshes so it didn't really simulate the issue being discussed.

I tried removing the animation part from class App and that reproduces the issue. Now, I see black screen when I toggle visibility. I tried applying the change from @michielvandergeest but that doesn't help.

In summary:
* The fix doesn't work on the test app (with animation part removed to simulate the black screen issue)
* Since the fix is in Lightning-SDK it won't help with apps which uses the Lightning core (like AppleTV+)

**Comment by varumugam123** [comment]
Hi @woutermeek @magomez @michielvandergeest Could you please review my comments and this workaround #903

**Comment by michielvandergeest** [comment]
@varumugam123 I have not been able to fully test the fix locally because I didn't have access to a device to test this on that displays this behaviour.

What I'd like to understand is what part of the proposed SDK solution doesn't work:

1) the visibility change call back doesn't run?
2) the callback runs, but the refocus, doesn't force a repaint

Would you be able to test this by adding a breakpoint (or console.log) to this part of the code: https://github.com/rdkcentral/Lightning-SDK/commit/6c37caab38da1f41c6cef1d063ca75ac572803ea

**Comment by varumugam123** [comment]
Hi @michielvandergeest, the callback is invoked and it even calls the refocus(), but then there is nothing on the screen...

![image](https://user-images.githubusercontent.com/29426674/177317774-64fc263c-c724-4d46-aaec-948878f9be24.png)

**Comment by magomez** [comment]
So it seems that the lightning fix is not valid because of the curious way the apps are using the framework. I'll have to fix it on the browser side then. Let's see if I can figure out a reliable way to do this without breaking the app's rendering.

**Comment by michielvandergeest** [comment]
@magomez there might be a different way to force a repaint. We'll be looking into that in the next days

**Comment by michielvandergeest** [comment]
@varumugam123 @magomez I've added a new commit to the fix-branch that might solve the problem: https://github.com/rdkcentral/Lightning-SDK/tree/fix/visibility-toggle-black-screen

Can you give it a try?

**Comment by frank-weindel** [comment]
@varumugam123 A version built into Lightning Core to try:
https://github.com/frank-weindel/Lightning/tree/feature/visibility-refresh

**Comment by varumugam123** [comment]
Hi @frank-weindel, WIth the above change I am able to see the content after visibility changes i.e (hidden -> visible)...

**Comment by woutermeek** [comment]
Glad that it got resolved, closing this issue now. Thanks @frank-weindel  @michielvandergeest  for the support.
